docs/design/features/source-generator-pinvokes.md

```diff
@@ -0,0 +1,232 @@
+
+``` CSharp
+/* A */ [DllImportAttribute("Kernel32.dll")]
+/* B */ extern static bool QueryPerformanceCounter(out long lpPerformanceCount);
+...
+long count;
+/* C */ QueryPerformanceCounter(out count);
+```
+
+At (A) in the above code snippet, the runtime is told to look for an export name `QueryPerformanceCounter` (B) in the `Kernel32.dll` binary. There are many additional attributes on the `DllImportAttribute` that help with export discovery and can influence the semantics of the generated IL Stub. Point (C) represents an invocation of the P/Invoke. Most of the work occurs at (C) at runtime, since (A) and (B) are merely declarations the compiler uses to embed the relevant details into assembly metadata that is read at runtime.
+
+1) During invocation, the function declaration is determined to be an external call requiring marshaling. Given the defined properties in the `DllImportAttribute` instance as well as the metadata of the user-defined signature an IL Stub is generated.
+
+2) The runtime attempts to find a binary with the name supplied in `DllImportAttribute`.
+
+    * Discovery of the target binary is complicated and can be influenced by the [`AssemblyLoadContext`](https://docs.microsoft.com/dotnet/api/system.runtime.loader.assemblyloadcontext) and [`NativeLibrary`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.nativelibrary) classes.
+
+3) Once the binary is found and loaded into the runtime, it is queried for the expected export name. The name of the attributed function is used by default but this is configurable by the [`DllImportAttribute.EntryPoint`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.dllimportattribute.entrypoint) property.
+
+    * This process is also influenced by additional `DllImportAttribute` properties as well as by the underlying platform. For example, the [Win32 API ANSI/UNICODE convention](https://docs.microsoft.com/windows/win32/intl/conventions-for-function-prototypes) on Windows is respected and a(n) `A`/`W` suffix may be appended to the function name if it is not immediately found.
+
+4) The IL Stub is called like any other .NET method.
+
+5) The IL Stub marshals arguments as appropriate and invokes the export via the `calli` instruction.
+
+6) Once the export returns control to the IL Stub the marshaling logic cleans up and ensures any returned data is marshaled back out to the calling function.
+
+### Source Generator Integration
+
+An example of how the previous P/Invoke snippet could be transformed is below. This example is using the proposed API in this document. The Source Generator currently has a limitation that we will ignore at present - the inability to modify user written code. This limitation is ignored to present an optimal user experience, but will be discussed later in case the current limitation becomes a hard requirement.
+
+`Program.cs` (Pre-generated code)
+
+``` CSharp
+/* A */ [DllImportAttribute("Kernel32.dll")]
```

Comment:
I don't like the idea that the code would have a use of DllImport attribute. It implies that this code might do something useful without the source generator attached. I'd much rather have the GeneratedDllImportAttribute have all of the various options that the source generator can handle. (Also by re-using DllImportAttribute, it requires that the initial implementation actually handle all of the capabilities of DllImport, or at least detect incorrect behaviors. That seems like a hard ask, and something that would drive a lot of unnecessary work to achieve an MVP.)

Reply:
>I don't like the idea that the code would have a use of DllImport attribute. It implies that this code might do something useful without the source generator attached. I'd much rather have the GeneratedDllImportAttribute have all of the various options that the source generator can handle.

I don't think that is an unreasonable perspective.

>That seems like a hard ask, and something that would drive a lot of unnecessary work to achieve an MVP.

Not sure I agree with this being apart of an MVP. MVPs don't technically require a bug free/comprehensive scenario state. Defining the supported path would be easy to define and stating the certain properties aren't considered at present would be enough for an MVP. The final product of course would warrant far more scrutiny.

Reply:
@davidwrighton I have adopted your feedback here and now the `GeneratedDllImportAttribute`  API will ideally have the same fields as `DllImportAttribute`. Pruning them is possible; suggestions are welcome.